### ISSUE TITLE
Explore integration of loss into trainings pipeline

### ISSUE THREAD

**Issue by michalharakal** [sk-ai-net/skainet#242]
### Summary

Define the forward loss-computation pipeline for SKaiNET that can be used immediately with existing models while remaining compatible with a future Trainer and training DSL.

---

## 🔁 Near-Term Pipeline

Provide a simple, explicit helper flow:

```kotlin
computeLoss(model, loss, x, y, ctx) {
    val logits = model.forward(x, ctx)
    return loss.forward(logits, y, ctx)
}
```

**Execution steps**

1. **Forward model**

   ```kotlin
   logits = model.forward(batchX, ctx)
   ```
2. **Compute loss**

   ```kotlin
   lossVal = loss.forward(logits, batchY, ctx)
   ```
3. **Return loss tensor**

   * Reduction applied inside loss (`NONE`, `MEAN`, or `SUM`)
   * Produces scalar tensor for `MEAN`/`SUM`, or per-sample tensor for `NONE`

---

## 🤖 Trainer Pipeline (Future)

Proposed training loop structure once gradient support exists:

```kotlin
logits = model.forward(batchX, ctx)
lossVal = loss.forward(logits, batchY, ctx)
backward(lossVal)
optimizer.step()
```

Configuration shape:

```kotlin
data class TrainingConfig(
    val loss: Loss,
    val reduction: Reduction = Reduction.MEAN
)
```

---

## 🛠 DSL Hook (Future)

High-level DSL tied to the same pipeline:

```kotlin
training {
    model { ... }
    loss { crossEntropy() }
    optimizer { ... }
}
```

Flow remains:

1. Model forward
2. Loss forward
3. Backprop
4. Optimizer step

---

## ✅ Acceptance Criteria

* A documented near-term usage pattern that executes:

  * model forward → loss forward → output tensor
* Pipeline is:

  * Explicit
  * Trainer-friendly
  * DSL-ready without coupling to module topology

**Comment by michalharakal** [comment]
Closing as implemented. Loss integration into the training pipeline ships in `nn/loss/Loss.kt` + `LossModule` (CE, MSE, BCE, ...).
